### PR TITLE
[onnx] Clarify that output_names labels outputs but does not reorder them

### LINKED ISSUE #165758: [ONNX] Better documentation needed for output_names arg
### 🐛 Describe the bug

Hi,

I've recently exported an encoder-decoder model to ONNX and found an issue regarding the order of the outputs when combined with the `output_names` parameter. When returning the KV cache in a dictionary, the returned values are in the order defined exclusively by the order of the return statement, regardless of the matching keys in the `output_names` parameter. I think this is an unexpected behavior that should be documented or changed.

Here is a minimal example:
```python
import onnxruntime as ort
import torch


class Model(torch.nn.Module):

    def forward(self, x):
        ones = torch.ones(1)
        zeros = torch.zeros(1)
        return {"zeros": zeros, "ones": ones}


model = Model()
x = torch.randn(1)
torch.onnx.export(
    model, (x), "model.onnx",
    input_names=["x"],
    output_names=["ones", "zeros"],
)

onnx_model = ort.InferenceSession("model.onnx")
ones, zeros = onnx_model.run(None, {"x": x.numpy()})

print("ones:", ones)
print("zeros:", zeros)

# ones: [0.]
# zeros: [1.]
```
The behavior I expected is that the `Model.forward` method does matching between the `output_names` parameter and the keys of the returned dictionary and returns them in the exact same order.

### Versions

```
Collecting environment information...
PyTorch version: 2.9.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version: (Debian 12.2.0-14) 12.2.0
Clang version: 14.0.6
CMake version: version 3.25.1
Libc version: glibc-2.36

Python version: 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.1.0-21-amd64-x86_64-with-glibc2.36
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.147.05
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 5950X 16-Core Processor
CPU family:                           25
Model:                                33
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             2
Frequency boost:                      enabled
CPU(s) scaling MHz:                   67%
CPU max MHz:                          5083.3979
CPU min MHz:                          2200.0000
BogoMIPS:                             6800.10
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                       AMD-V
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             8 MiB (16 instances)
L3 cache:                             64 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.3.4
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.19.1
[pip3] onnx-ir==0.1.11
[pip3] onnxruntime==1.23.1
[pip3] onnxscript==0.5.4
[pip3] torch==2.9.0
[pip3] triton==3.5.0
[conda] Could not collect
 ```

cc @justinchuby @titaiwangms @svekars @sekyondaMeta @AlannaBurke

### PR DESCRIPTION
**Pull request by JStoschek**
Fixes #165758

## Description

`output_names` in `torch.onnx.export` only assigns labels to output nodes. It
does not control their order. When a model returns a dictionary, outputs are
flattened in the dictionary's iteration order regardless of what names are
specified. This was not documented, leading to user confusion (see linked issue).

This PR adds a clarifying note to the `output_names` parameter docstring in
both the dynamo-based exporter (`torch/onnx/__init__.py`) and the legacy
TorchScript exporter (`torch/onnx/_internal/torchscript_exporter/utils.py`).

cc @justinchuby @titaiwangms

### CODE REVIEW

**Review by justinchuby — DISMISSED** [review]

**Comment by JStoschek** [comment]
Hi @justinchuby, this PR got closed by the stale bot while I was heads down on other things, but it looks like it was already approved and rebased on main. Could this be reopened and merged? I noticed someone opened #188893 reviving the same fix in the meantime, so happy to coordinate however makes sense on your end, but would prefer to get this one landed directly if possible since it's ready to go. Thanks for the review back in April, and sorry for letting it go stale.

**Review by justinchuby — APPROVED** [review]
